atc/exec/on_error.go

```diff
@@ -0,0 +1,47 @@
+package exec
+
+import (
+	"context"
+)
+
+// OnErrorStep will run one step, and then a second step if the first step
+// errors.
+type OnErrorStep struct {
+	step Step
+	hook Step
+}
+
+// OnError constructs an OnErrorStep factory.
+func OnError(step Step, hook Step) OnErrorStep {
+	return OnErrorStep{
+		step: step,
+		hook: hook,
+	}
+}
+
+// Run will call Run on the first step and wait for it to complete. If the
+// first step errors, Run returns the error. OnErrorStep is ready as soon as
+// the first step is ready.
+//
+// If the first step errors, the second
+// step is executed. If the second step errors, nothing is returned.
+func (o OnErrorStep) Run(ctx context.Context, state RunState) error {
+	stepRunErr := o.step.Run(ctx, state)
+	// with no error, we just want to return right away
+	if stepRunErr == nil {
+		return nil
+	}
+
+	// for all errors that aren't caused by an Abort, run the hook
+	if stepRunErr != context.Canceled {
+		o.hook.Run(context.Background(), state)
```

Comment:
Might want an explicit `_ =` here to make it clear that we're intentionally dropping the error.

Do you think it'd be interesting to return both errors if the error hook errors? That could be done with `multierror`. I have no real preference.

Reply:
Good point on the multierror. Will add that in.